feat: Add support for macOS layered icons

Sanlorng opened the pull request:
Support for using macOS layered icons (`.icon` directory) for both JVM and native application packaging.
> This feature requires actool version >= 26.0

- A new `layeredIconDir` property is added to the `compose.desktop.application.nativeDistributions.macOS` and `compose.nativeApplication.macOS` DSLs.
- The `actool` from Xcode Command Line Tools is used to compile the `.icon` directory into an `Assets.car` file.
- The `Assets.car` file is included in the `.app` bundle resources.
- The `Info.plist` is updated with `CFBundleIconName` to reference the compiled asset.
- A check for a supported `actool` version is performed before compilation.
- New integration tests (`testMacLayeredIcon`, `testMacLayeredIconRemove`) are added to verify the functionality for both JVM and native targets.

<img width="144" height="132" alt="image" src="https://github.com/user-attachments/assets/96ec9056-d629-4dbf-8457-18f151a3d963" />
<img width="142" height="136" alt="image" src="https://github.com/user-attachments/assets/70d433ae-627b-4f7f-9382-31ca8c412237" />
<img width="142" height="132" alt="image" src="https://github.com/user-attachments/assets/262d864e-da01-428b-b676-2f01dbcf86c6" />
<img width="142" height="140" alt="image" src="https://github.com/user-attachments/assets/0e8bb6b8-01c3-4f22-bb07-7945f3624a48" />



## Example
```kotlin
compose.desktop {
    application {
        nativeDistributions {
            macOS {
                layeredIconDir.set("Icon Path")
            }
        }
    }
}
```


## Testing
**Only verify the task behavior**: compose:test-Gradle(9.2.0)-Agp(8.12.3) --tests org.jetbrains.compose.test.tests.integration.DesktopApplicationTest.testMacLayeredIcon

**Only verify the task behavior**: compose:test-Gradle(9.2.0)-Agp(8.12.3) --tests org.jetbrains.compose.test.tests.integration.DesktopApplicationTest.testMacLayeredIconRemove

**This should be tested by QA**

## Tickets
[CMP-9083](https://youtrack.jetbrains.com/issue/CMP-9083) Support macOS 26 layered icons

## Release Notes
### Features - Gradle Plugin
- Support for using macOS layered icons (`.icon` directory) for both JVM and native application packaging.

Sanlorng committed:
feat: Add support for macOS layered icons

This commit introduces support for using macOS layered icons (`.icon` directory) for both JVM and native application packaging.

- A new `layeredIconDir` property is added to the `compose.desktop.application.nativeDistributions.macOS` and `compose.nativeApplication.macOS` DSLs.
- The `actool` from Xcode Command Line Tools is used to compile the `.icon` directory into an `Assets.car` file.
- The `Assets.car` file is included in the `.app` bundle resources.
- The `Info.plist` is updated with `CFBundleIconName` to reference the compiled asset.
- A check for a supported `actool` version is performed before compilation.
- New integration tests (`testMacLayeredIcon`, `testMacLayeredIconRemove`) are added to verify the functionality for both JVM and native targets.

Sanlorng committed:
Remove unused `macAssetsDir` property

The `macAssetsDir` property in `AbstractJPackageTask` was unused and has been removed.

Sanlorng committed:
clean code

- Deleted the `Expected-Info.plist` which verified now-removed properties.
- Removed `extraInfoPlistKeys` and file association configurations from the test's `build.gradle.kts`.

Sanlorng commented:
Thanks again for creating the related YouTrack issue – it’s been really helpful!

I just wanted to gently follow up on this PR – would you have a moment to review it when you get a chance? Let me know if I can clarify anything to make it easier!

No rush at all, appreciate your help!

freelife4850-netizen reviewed (commented):
Sounds like a good idea to try 🙂

kropp commented on gradle-plugins/compose/src/test/kotlin/org/jetbrains/compose/test/tests/integration/DesktopApplicationTest.kt:
```diff
@@ -366,6 +367,77 @@ class DesktopApplicationTest : GradlePluginTestBase() {
 
+    @Test
+    fun testMacLayeredIcon() {
+        Assumptions.assumeTrue(currentOS == OS.MacOS)
```
I suppose that these tests will only run with SDK 26, please add an assumption for that as well.

kropp reviewed (changes requested):
Sorry for a long wait, could you please take a look at comments and also rebase the branch to the latest version?

Sanlorng committed:
Refactor: Use placeholders for Compose dependencies in test

Updated the `macLayeredIcon` test project to use `COMPOSE_VERSION_PLACEHOLDER` for Compose dependency versions. This makes the test configuration more flexible and easier to maintain.

Sanlorng committed:
Refactor: Use `plutil` to parse `actool` version

This commit refactors the `actool` version parsing logic to use the `plutil` command-line tool instead of a manual XML parser. This change simplifies the code and improves robustness.

- Replaced the XML parsing implementation in `MacAssetsTool` with a call to `plutil -extract`.
- Updated `try`/`catch` blocks to handle exceptions from the tool execution.
- Added a `plutil` file definition to `MacUtils`.
- Integration tests for macOS layered icons (`testMacLayeredIcon`, `testMacLayeredIconRemove`) are now skipped if the `actool` version is less than 26, ensuring they run only on supported environments.

kropp commented:
@igordmn requesting second review as it adds new public API in Gradle plugin

PierreVieira commented on gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/internal/MacAssetsTool.kt:
```diff
@@ -0,0 +1,99 @@
+            return versionString
+        }
+    }
+}
```
Missing EOF here


```suggestion
}

```

PierreVieira commented on gradle-plugins/compose/src/test/test-projects/application/macLayeredIcon/src/macosMain/kotlin/main.kt:
```diff
@@ -0,0 +1,4 @@
+
+fun main() {
+    println("Hello, from Mac OS!")
+}
```
```suggestion
}

```

PierreVieira commented on gradle-plugins/compose/src/test/test-projects/application/macLayeredIcon/build.gradle.kts:
```diff
@@ -0,0 +1,84 @@
+            }
+        }
+    }
+}
```
```suggestion
}

```

PierreVieira reviewed (commented):
Simple comments. Nice addition!

Sanlorng committed:
typo: add new lines

Co-authored-by: Pierre Vieira <pierrevieiraggg@gmail.com>

igordmn reviewed (changes requested):
Thank you and sorry for more waiting time. After we resolve the new comments, it should be good to merge.

Sanlorng commented:
> Sanlorng, please confirm that you still plan to work on it.
> 
> If not, we will try to finish it and merge it later.

I may not be able to continue testing on my Mac mini. After rebasing to the April master branch, compiling locally keeps failing with a Java heap dump OOM error. You might need to handle the remaining code changes from here.
